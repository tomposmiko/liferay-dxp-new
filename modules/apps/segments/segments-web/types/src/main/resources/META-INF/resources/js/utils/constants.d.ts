/**
 * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
 *
 * This library is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Lesser General Public License as published by the Free
 * Software Foundation; either version 2.1 of the License, or (at your option)
 * any later version.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
 * details.
 */

export declare const CONJUNCTIONS: {
	readonly AND: 'and';
	readonly OR: 'or';
};
export declare const FUNCTIONAL_OPERATORS: {
	readonly CONTAINS: 'contains';
};
export declare const NOT_OPERATORS: {
	readonly NOT_CONTAINS: 'not-contains';
	readonly NOT_EQ: 'not-eq';
};
export declare const HAS_OPERATORS: {
	readonly HAS: 'has';
	readonly NOT_HAS: 'has-not';
};
export declare const RELATIONAL_OPERATORS: {
	readonly EQ: 'eq';
	readonly GE: 'ge';
	readonly GT: 'gt';
	readonly LE: 'le';
	readonly LT: 'lt';
};
export declare const DATE_OPERATORS: {
	readonly BETWEEN: 'between';
	readonly EVER: 'ever';
};
export declare const SINCE_VALUES: {
	readonly last7Days: string;
	readonly last24Hours: string;
	readonly last28Days: string;
	readonly last30Days: string;
	readonly last90Days: string;
	readonly yesterday: string;
};

/**
 * Constants to match property types in the passed in supportedProperties array.
 */
export declare const PROPERTY_TYPES: {
	readonly BOOLEAN: 'boolean';
	readonly COLLECTION: 'collection';
	readonly DATE: 'date';
	readonly DATE_TIME: 'date-time';
	readonly DOUBLE: 'double';
	readonly EVENT: 'event';
	readonly ID: 'id';
	readonly INTEGER: 'integer';
	readonly STRING: 'string';
};
export declare const PROPERTY_GROUPS: {
	readonly EVENT: 'event';
};
export declare const SUPPORTED_CONJUNCTIONS: readonly [
	{
		readonly label: string;
		readonly name: 'and';
	},
	{
		readonly label: string;
		readonly name: 'or';
	}
];
export declare const SUPPORTED_OPERATORS: readonly [
	{
		readonly label: string;
		readonly name: 'eq';
	},
	{
		readonly label: string;
		readonly name: 'not-eq';
	},
	{
		readonly label: string;
		readonly name: 'gt';
	},
	{
		readonly label: string;
		readonly name: 'ge';
	},
	{
		readonly label: string;
		readonly name: 'lt';
	},
	{
		readonly label: string;
		readonly name: 'le';
	},
	{
		readonly label: string;
		readonly name: 'contains';
	},
	{
		readonly label: string;
		readonly name: 'not-contains';
	}
];
export declare const SUPPORTED_EVENT_OPERATORS: readonly [
	{
		readonly label: string;
		readonly name: 'ge';
	},
	{
		readonly label: string;
		readonly name: 'le';
	},
	{
		readonly label: string;
		readonly name: 'has';
	},
	{
		readonly label: string;
		readonly name: 'has-not';
	}
];
export declare const SUPPORTED_EVENT_DATE_OPERATORS: readonly [
	{
		readonly label: string;
		readonly name: 'between';
	},
	{
		readonly label: string;
		readonly name: 'eq';
	},
	{
		readonly label: string;
		readonly name: 'ever';
	},
	{
		readonly label: string;
		readonly name: 'gt';
	},
	{
		readonly label: string;
		readonly name: 'lt';
	}
];
export declare const SUPPORTED_PROPERTY_TYPES: {
	readonly 'boolean': readonly ['eq', 'not-eq'];
	readonly 'collection': readonly [
		'eq',
		'not-eq',
		'contains',
		'not-contains'
	];
	readonly 'date': readonly ['eq', 'ge', 'gt', 'le', 'lt', 'not-eq'];
	readonly 'date-time': readonly ['eq', 'ge', 'gt', 'le', 'lt', 'not-eq'];
	readonly 'double': readonly ['eq', 'ge', 'gt', 'le', 'lt', 'not-eq'];
	readonly 'event': {
		readonly DATE: readonly ['between', 'eq', 'ever', 'gt', 'lt'];
		readonly INTEGER: readonly ['ge', 'le'];
		readonly NOT: readonly ['has', 'has-not'];
		readonly SINCE: readonly [
			string,
			string,
			string,
			string,
			string,
			string
		];
	};
	readonly 'id': readonly ['eq', 'not-eq'];
	readonly 'integer': readonly ['eq', 'ge', 'gt', 'le', 'lt', 'not-eq'];
	readonly 'string': readonly ['eq', 'not-eq', 'contains', 'not-contains'];
};

/**
 * Values for criteria row inputs.
 */
export declare const BOOLEAN_OPTIONS: readonly [
	{
		readonly label: string;
		readonly value: 'true';
	},
	{
		readonly label: string;
		readonly value: 'false';
	}
];
