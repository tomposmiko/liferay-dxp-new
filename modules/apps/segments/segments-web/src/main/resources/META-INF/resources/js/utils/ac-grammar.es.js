/**
 * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
 *
 * This library is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Lesser General Public License as published by the Free
 * Software Foundation; either version 2.1 of the License, or (at your option)
 * any later version.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
 * details.
 */

/**
 * Recursively traverses the criteria object to build an Ac Gramar filter query
 * string. Properties is required to parse the correctly with or without quotes
 * and formatting the query differently for certain types like collection.
 * @returns An AC grammar query string built from the criteria object.
 */
function buildEventQueryString(criteria) {
	const {items} = criteria;
	const operatorName = items[0]?.operatorName;
	const value = items[0]?.value;

	let query = `activities.filterByCount(filter='()',operator='${operatorName}',value=${value})`;

	query = items[0]?.operatorNot ? `((not ${query}))` : `(${query})`;

	return query;
}

export {buildEventQueryString};
