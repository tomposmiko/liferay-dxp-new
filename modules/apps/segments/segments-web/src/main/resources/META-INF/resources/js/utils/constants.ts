/**
 * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
 *
 * This library is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Lesser General Public License as published by the Free
 * Software Foundation; either version 2.1 of the License, or (at your option)
 * any later version.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
 * details.
 */

/**
 * Constants for OData query.
 */

import {sub} from 'frontend-js-web';

export const CONJUNCTIONS = {
	AND: 'and',
	OR: 'or',
} as const;

export const FUNCTIONAL_OPERATORS = {
	CONTAINS: 'contains',
} as const;

export const NOT_OPERATORS = {
	NOT_CONTAINS: 'not-contains',
	NOT_EQ: 'not-eq',
} as const;

export const HAS_OPERATORS = {
	HAS: 'has',
	NOT_HAS: 'has-not',
} as const;

export const RELATIONAL_OPERATORS = {
	EQ: 'eq',
	GE: 'ge',
	GT: 'gt',
	LE: 'le',
	LT: 'lt',
} as const;

export const DATE_OPERATORS = {
	BETWEEN: 'between',
	EVER: 'ever',
} as const;

export const SINCE_VALUES = {
	last7Days: sub(Liferay.Language.get('last-x-days'), 7),
	last24Hours: sub(Liferay.Language.get('last-x-hours'), 24),
	last28Days: sub(Liferay.Language.get('last-x-days'), 28),
	last30Days: sub(Liferay.Language.get('last-x-days'), 30),
	last90Days: sub(Liferay.Language.get('last-x-days'), 90),
	yesterday: Liferay.Language.get('yesterday'),
} as const;

/**
 * Constants to match property types in the passed in supportedProperties array.
 */

export const PROPERTY_TYPES = {
	BOOLEAN: 'boolean',
	COLLECTION: 'collection',
	DATE: 'date',
	DATE_TIME: 'date-time',
	DOUBLE: 'double',
	EVENT: 'event',
	ID: 'id',
	INTEGER: 'integer',
	STRING: 'string',
} as const;

export const PROPERTY_GROUPS = {
	EVENT: 'event',
} as const;

/**
 * Constants for CriteriaBuilder component.
 */

const {AND, OR} = CONJUNCTIONS;
const {EQ, GE, GT, LE, LT} = RELATIONAL_OPERATORS;
const {NOT_CONTAINS, NOT_EQ} = NOT_OPERATORS;
const {HAS, NOT_HAS} = HAS_OPERATORS;
const {CONTAINS} = FUNCTIONAL_OPERATORS;
const {
	BOOLEAN,
	COLLECTION,
	DATE,
	DATE_TIME,
	DOUBLE,
	EVENT,
	ID,
	INTEGER,
	STRING,
} = PROPERTY_TYPES;
const {BETWEEN, EVER} = DATE_OPERATORS;
const {
	last7Days,
	last24Hours,
	last28Days,
	last30Days,
	last90Days,
	yesterday,
} = SINCE_VALUES;

export const SUPPORTED_CONJUNCTIONS = [
	{
		label: Liferay.Language.get('and'),
		name: AND,
	},
	{
		label: Liferay.Language.get('or'),
		name: OR,
	},
] as const;

export const SUPPORTED_OPERATORS = [
	{
		label: Liferay.Language.get('equals'),
		name: EQ,
	},
	{
		label: Liferay.Language.get('not-equals'),
		name: NOT_EQ,
	},
	{
		label: Liferay.Language.get('greater-than'),
		name: GT,
	},
	{
		label: Liferay.Language.get('greater-than-or-equals'),
		name: GE,
	},
	{
		label: Liferay.Language.get('less-than'),
		name: LT,
	},
	{
		label: Liferay.Language.get('less-than-or-equals'),
		name: LE,
	},
	{
		label: Liferay.Language.get('contains'),
		name: CONTAINS,
	},
	{
		label: Liferay.Language.get('not-contains'),
		name: NOT_CONTAINS,
	},
] as const;

export const SUPPORTED_EVENT_OPERATORS = [
	{
		label: Liferay.Language.get('at-least'),
		name: GE,
	},
	{
		label: Liferay.Language.get('at-most'),
		name: LE,
	},
	{
		label: Liferay.Language.get('has'),
		name: HAS,
	},
	{
		label: Liferay.Language.get('has-not'),
		name: NOT_HAS,
	},
] as const;

export const SUPPORTED_EVENT_DATE_OPERATORS = [
	{
		label: Liferay.Language.get('between'),
		name: BETWEEN,
	},
	{
		label: Liferay.Language.get('on'),
		name: EQ,
	},
	{
		label: Liferay.Language.get('ever'),
		name: EVER,
	},
	{
		label: Liferay.Language.get('after'),
		name: GT,
	},
	{
		label: Liferay.Language.get('before'),
		name: LT,
	},
] as const;

export const SUPPORTED_PROPERTY_TYPES = {
	[BOOLEAN]: [EQ, NOT_EQ],
	[COLLECTION]: [EQ, NOT_EQ, CONTAINS, NOT_CONTAINS],
	[DATE]: [EQ, GE, GT, LE, LT, NOT_EQ],
	[DATE_TIME]: [EQ, GE, GT, LE, LT, NOT_EQ],
	[DOUBLE]: [EQ, GE, GT, LE, LT, NOT_EQ],
	[EVENT]: {
		DATE: [BETWEEN, EQ, EVER, GT, LT],
		INTEGER: [GE, LE],
		NOT: [HAS, NOT_HAS],
		SINCE: [
			last7Days,
			last24Hours,
			last28Days,
			last30Days,
			last90Days,
			yesterday,
		],
	},
	[ID]: [EQ, NOT_EQ],
	[INTEGER]: [EQ, GE, GT, LE, LT, NOT_EQ],
	[STRING]: [EQ, NOT_EQ, CONTAINS, NOT_CONTAINS],
} as const;

/**
 * Values for criteria row inputs.
 */

export const BOOLEAN_OPTIONS = [
	{
		label: Liferay.Language.get('true'),
		value: 'true',
	},
	{
		label: Liferay.Language.get('false'),
		value: 'false',
	},
] as const;
