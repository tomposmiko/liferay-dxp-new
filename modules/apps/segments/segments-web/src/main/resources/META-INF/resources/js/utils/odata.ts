/**
 * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
 *
 * This library is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Lesser General Public License as published by the Free
 * Software Foundation; either version 2.1 of the License, or (at your option)
 * any later version.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
 * details.
 */

import {Criteria, CriteriaItem, Property} from '../../types/Criteria';
import {
	Conjunction,
	FUNCTIONAL_OPERATORS,
	NOT_OPERATORS,
	PROPERTY_TYPES,
	PropertyType,
	RELATIONAL_OPERATORS,
} from './constants';

/**
 * Gets the type of the property from the property name.
 */
const getTypeByPropertyName = (
	propertyName: string,
	properties: Array<{name: string; type: PropertyType}>
) => {
	let type: PropertyType | undefined;

	if (propertyName && properties) {
		const property = properties.find(
			(property) => property.name === propertyName
		);

		if (property) {
			type = property.type;
		}
	}

	return type;
};

/**
 * Handles the single quotes present in the value.
 * Should escape the single quotes present in the value before hit the backend.
 */
function escapeSingleQuotes<T>(value: T) {
	if (typeof value !== 'string') {
		return value;
	}

	return `'${value.replace(/'/g, "''")}'`;
}

/**
 * Handles the single quotes present in the value.
 * Should un-escape the single quotes present in odata format before rendering.
 */
function unescapeSingleQuotes<T>(value: T) {
	if (typeof value !== 'string') {
		return value;
	}

	return value.replace(/''/g, "'");
}

/**
 * Decides whether to add quotes to value.
 */
function valueParser(value: unknown, type: PropertyType | undefined) {
	let parsedValue;

	switch (type) {
		case PROPERTY_TYPES.BOOLEAN:
		case PROPERTY_TYPES.DATE:
		case PROPERTY_TYPES.DATE_TIME:
		case PROPERTY_TYPES.INTEGER:
		case PROPERTY_TYPES.DOUBLE:
			parsedValue = value;
			break;
		case PROPERTY_TYPES.COLLECTION:
		case PROPERTY_TYPES.STRING:
		default:
			parsedValue = escapeSingleQuotes(value);
			break;
	}

	return parsedValue;
}

/**
 * Recursively traverses the criteria object to build an oData filter query
 * string. Properties is required to parse the correctly with or without quotes
 * and formatting the query differently for certain types like collection.
 * @param {object} criteria The criteria object.
 * @param {string} queryConjunction The conjunction name value to be used in the
 * query.
 * @param {array} properties The list of property objects. See
 * ContributorBuilder for valid property object shape.
 * @returns An OData query string built from the criteria object.
 */
function buildQueryString(
	criteria: Array<Criteria | CriteriaItem | null>,
	queryConjunction: Conjunction,
	properties: Property[]
) {
	return (criteria.filter(Boolean) as Array<Criteria | CriteriaItem>).reduce(
		(queryString, criteriaOrCriteriaItem, index) => {
			if (index > 0) {
				queryString = queryString.concat(` ${queryConjunction} `);
			}

			if ('conjunctionName' in criteriaOrCriteriaItem) {
				queryString = queryString.concat(
					`(${buildQueryString(
						criteriaOrCriteriaItem.items,
						criteriaOrCriteriaItem.conjunctionName,
						properties
					)})`
				);
			}
			else {
				const {
					operatorName,
					propertyName,
					value,
				} = criteriaOrCriteriaItem;

				const type =
					criteriaOrCriteriaItem.type ||
					getTypeByPropertyName(propertyName, properties);

				const parsedValue = valueParser(value, type);

				if (isValueType(RELATIONAL_OPERATORS, operatorName)) {
					if (type === PROPERTY_TYPES.COLLECTION) {
						queryString = queryString.concat(
							`${propertyName}/any(c:c ${operatorName} ${parsedValue})`
						);
					}
					else {
						queryString = queryString.concat(
							`${propertyName} ${operatorName} ${parsedValue}`
						);
					}
				}
				else if (isValueType(FUNCTIONAL_OPERATORS, operatorName)) {
					if (type === PROPERTY_TYPES.COLLECTION) {
						queryString = queryString.concat(
							`${propertyName}/any(c:${operatorName}(c, ${parsedValue}))`
						);
					}
					else {
						queryString = queryString.concat(
							`${operatorName}(${propertyName}, ${parsedValue})`
						);
					}
				}
				else if (isValueType(NOT_OPERATORS, operatorName)) {
					const baseOperator = operatorName.replace(/not-/g, '');

					const baseExpression: CriteriaItem[] = [
						{
							operatorName: baseOperator,
							propertyName,
							type,
							value,
						},
					];

					// Not is wrapped in a group to simplify AST parsing.

					queryString = queryString.concat(
						`(not (${buildQueryString(
							baseExpression,
							queryConjunction,
							properties
						)}))`
					);
				}
			}

			return queryString;
		},
		''
	);
}

/**
 * Checks if the value is a certain type.
 */
function isValueType<Map extends Record<K, V>, K extends string, V>(
	types: Map,
	value: unknown
): boolean {
	return Object.values(types).includes(value);
}

export {buildQueryString, unescapeSingleQuotes};
