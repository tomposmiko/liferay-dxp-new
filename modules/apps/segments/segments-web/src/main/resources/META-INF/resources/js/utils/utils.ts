/**
 * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
 *
 * This library is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Lesser General Public License as published by the Free
 * Software Foundation; either version 2.1 of the License, or (at your option)
 * any later version.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
 * details.
 */

import {format, isValid, parseISO} from 'date-fns';

import {Criteria} from '../../types/Criteria';
import {CONJUNCTIONS, PropertyType} from './constants';

const GROUP_ID_NAMESPACE = 'group_';

const SPLIT_REGEX = /({\d+})/g;

/**
 * Creates a new group object with items.
 */
export function createNewGroup(items: Criteria['items']): Criteria {
	return {
		conjunctionName: CONJUNCTIONS.AND,
		groupId: generateGroupId(),
		items,
	};
}

let uniqueIdCounter_ = 1;

/**
 * Generates a unique group id.
 */
export function generateGroupId() {
	return `${GROUP_ID_NAMESPACE}${uniqueIdCounter_++}`;
}

/**
 * Uses the singular language key if the count is 1. Otherwise uses the plural
 * language key.
 */
export function getPluralMessage(
	singular: string,
	plural: string,
	count: number = 0,
	toString: boolean = false
) {
	const message = count === 1 ? singular : plural;

	return sub(message, [count.toString()], toString);
}

/**
 * Gets a list of group ids from a criteria object.
 * Used for disallowing groups to be moved into its own deeper nested groups.
 * Example of returned value: ['group_02', 'group_03']
 */
export function getChildGroupIds(criteria: Criteria) {
	let childGroupIds: string[] = [];

	if (criteria.items && criteria.items.length) {
		childGroupIds = criteria.items.reduce((groupIdList: string[], item) => {
			return 'groupId' in item
				? [...groupIdList, item.groupId, ...getChildGroupIds(item)]
				: groupIdList;
		}, []);
	}

	return childGroupIds;
}

/**
 * Gets the list of operators for a supported type.
 * Used for displaying the operators available for each criteria row.
 */
export function getSupportedOperatorsFromType<
	Operator extends {name: string},
	PropertyKey extends string
>(
	operators: Operator[],
	propertyTypes: Record<PropertyKey, PropertyType>,
	type: PropertyKey
) {
	return operators.filter((operator) => {
		const validOperators = propertyTypes[type];

		return validOperators && validOperators.includes(operator.name);
	});
}

/**
 * Inserts an item into a list at the specified index.
 */
export function insertAtIndex<T>(item: T, list: T[], index: number) {
	return [...list.slice(0, index), item, ...list.slice(index, list.length)];
}

/**
 * Converts an object of key value pairs to a form data object for passing
 * into a fetch body.
 */
export function objectToFormData(dataObject: Record<string, string | Blob>) {
	const formData = new FormData();

	Object.keys(dataObject).forEach((key) => {
		formData.set(key, dataObject[key]);
	});

	return formData;
}

/**
 * Removes an item at the specified index.
 */
export function removeAtIndex<T>(list: T[], index: number) {
	return list.filter((fItem, fIndex) => fIndex !== index);
}

/**
 * Replaces an item in a list at the specified index.
 */
export function replaceAtIndex<T>(item: T, list: T[], index: number) {
	return Object.assign(list, {
		[index]: item,
	});
}

/**
 * Utility function for substituting variables into language keys.
 *
 * Examples:
 * sub(Liferay.Language.get('search-x'), ['all'])
 * => 'search all'
 * sub(Liferay.Language.get('search-x'), [<b>all<b>], false)
 * => 'search <b>all</b>'
 *
 * Join boolean is used to indicate whether to call `.join()` on
 * the array before it is returned. Use `false` if subbing in JSX.
 */
export function sub(langKey: string, args: string[], join = true) {
	const keyArray = langKey
		.split(SPLIT_REGEX)
		.filter((val) => val.length !== 0);

	for (let i = 0; i < args.length; i++) {
		const arg = args[i];

		const indexKey = `{${i}}`;

		let argIndex = keyArray.indexOf(indexKey);

		while (argIndex >= 0) {
			keyArray.splice(argIndex, 1, arg);

			argIndex = keyArray.indexOf(indexKey);
		}
	}

	return join ? keyArray.join('') : keyArray;
}

export function dateToInternationalHuman(
	ISOString: string,
	localeKey = Liferay.ThemeDisplay.getBCP47LanguageId()
) {
	const date = new Date(ISOString);

	const intl = new Intl.DateTimeFormat(localeKey, {
		day: 'numeric',
		month: 'long',
		year: 'numeric',
	});

	return intl.format(date);
}

/**
 * Returns a YYYY-MM-DD date
 * based on a JS Date object
 */
export function jsDatetoYYYYMMDD(dateJsObject: string | Date) {
	if (!isValid(dateJsObject)) {
		dateJsObject = parseISO(dateJsObject as string);
	}

	return format(dateJsObject as Date, 'yyyy-MM-dd');
}
