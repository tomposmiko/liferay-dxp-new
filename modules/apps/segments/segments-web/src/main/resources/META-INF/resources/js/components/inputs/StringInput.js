/**
 * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
 *
 * This library is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Lesser General Public License as published by the Free
 * Software Foundation; either version 2.1 of the License, or (at your option)
 * any later version.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
 * details.
 */

import {ClayInput, ClaySelectWithOption} from '@clayui/form';
import classNames from 'classnames';
import propTypes from 'prop-types';
import React, {useState} from 'react';

import {unescapeSingleQuotes} from '../../utils/odata';

function StringInput({
	disabled,
	onChange,
	options = [],
	propertyLabel,
	renderEmptyValueErrors,
	value: initialValue,
}) {
	const [value, setValue] = useState(unescapeSingleQuotes(initialValue));

	const handleChange = (event) => {
		setValue(event.target.value);

		onChange({value: event.target.value});
	};

	return !options.length ? (
		<ClayInput
			aria-label={`${propertyLabel}: ${Liferay.Language.get(
				'input-a-value'
			)}`}
			className={classNames('criterion-input', {
				'criterion-input--error': !value && renderEmptyValueErrors,
			})}
			data-testid="simple-string"
			disabled={disabled}
			onChange={handleChange}
			type="text"
			value={value}
		/>
	) : (
		<ClaySelectWithOption
			aria-label={`${propertyLabel}: ${Liferay.Language.get(
				'select-option'
			)}`}
			className={classNames('criterion-input', {
				'criterion-input--error': !value,
			})}
			data-testid="options-string"
			disabled={disabled}
			onChange={handleChange}
			options={options.map((o) => ({
				disabled: o.disabled,
				label: o.label,
				value: o.value,
			}))}
			value={value}
		/>
	);
}

StringInput.propTypes = {
	disabled: propTypes.bool,
	initialValue: propTypes.oneOfType([propTypes.string, propTypes.number]),
	onChange: propTypes.func.isRequired,
	options: propTypes.array,
	propertyLabel: propTypes.string.isRequired,
	renderEmptyValueErrors: propTypes.bool,
};

export default StringInput;
