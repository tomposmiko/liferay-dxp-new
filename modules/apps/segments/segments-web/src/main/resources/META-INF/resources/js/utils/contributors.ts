/**
 * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
 *
 * This library is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Lesser General Public License as published by the Free
 * Software Foundation; either version 2.1 of the License, or (at your option)
 * any later version.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
 * details.
 */

import {Property} from '../../types/Criteria';
import {CONJUNCTIONS, Conjunction, SUPPORTED_CONJUNCTIONS} from './constants';
import {buildQueryString} from './odata';

interface Contributor {
	conjunctionId: Conjunction;
	conjunctionInputId: string;
	criteriaMap: Record<string, unknown>;
	entityName: string;
	initialQuery?: Parameters<typeof buildQueryString>[0][0];
	inputId: string;
	modelLabel: string;
	properties: Property[];
	propertyKey: unknown;
	query: string;
}

interface PropertyGroup {
	entityName: string;
	name: string;
	properties: Property[];
	propertyKey: string;
}

/**
 * Produces a list of Contributors from a list of initialContributors
 * and a list of propertyGroups.
 */
export function initialContributorsToContributors(
	initialContributors: Contributor[],
	propertyGroups: PropertyGroup[]
) {
	const DEFAULT_CONTRIBUTOR = {conjunctionId: CONJUNCTIONS.AND};
	const {conjunctionId: initialConjunction} =
		initialContributors.find((c) => c.conjunctionId) || DEFAULT_CONTRIBUTOR;

	return initialContributors.map((initialContributor) => {
		const propertyGroup =
			propertyGroups &&
			propertyGroups.find(
				(propertyGroup) =>
					initialContributor.propertyKey === propertyGroup.propertyKey
			);

		return {
			conjunctionId:
				initialContributor.conjunctionId || initialConjunction,
			conjunctionInputId: initialContributor.conjunctionInputId,
			criteriaMap: initialContributor.initialQuery || null,
			entityName: propertyGroup && propertyGroup.entityName,
			inputId: initialContributor.inputId,
			modelLabel: propertyGroup && propertyGroup.name,
			properties: propertyGroup && propertyGroup.properties,
			propertyKey: initialContributor.propertyKey,
			query: initialContributor.initialQuery
				? buildQueryString(
						[initialContributor.initialQuery],
						initialContributor.conjunctionId || initialConjunction,
						propertyGroup?.properties || []
				  )
				: '',
		};
	});
}

/**
 * Applies a criteria change to a contributor from a list in both the
 * criteriaMap and query properties.
 */
export function applyCriteriaChangeToContributors(
	contributors: Contributor[],
	change: {
		criteriaChange: Contributor['initialQuery'];
		propertyKey: PropertyKey;
	}
) {
	return contributors.map((contributor) => {
		const {conjunctionId, properties, propertyKey} = contributor;

		return change.propertyKey === propertyKey
			? {
					...contributor,
					criteriaMap: change.criteriaChange,
					query: buildQueryString(
						[change.criteriaChange || null],
						conjunctionId,
						properties
					),
			  }
			: contributor;
	});
}

/**
 * Applies a conjunction change to the whole array of contributors.
 */
export function applyConjunctionChangeToContributor(
	contributors: Contributor[],
	conjunctionName: Conjunction
) {
	const conjunctionIndex = SUPPORTED_CONJUNCTIONS.findIndex(
		(item) => item.name === conjunctionName
	);

	if (conjunctionIndex === -1) {
		return contributors;
	}

	const nextContributors = contributors.map((contributor) => ({
		...contributor,
		conjunctionId: conjunctionName,
	}));

	return nextContributors;
}
