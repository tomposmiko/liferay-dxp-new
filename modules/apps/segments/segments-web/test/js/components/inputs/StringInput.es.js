import React from 'react';
import StringInput from 'components/inputs/StringInput.es';
import {cleanup, render} from 'react-testing-library';
import {testControlledInput} from 'test/utils';

const OPTIONS_STRING_INPUT_TESTID = 'options-string';

const SIMPLE_STRING_INPUT_TESTID = 'simple-string';

const defaultValue = 'defaultValue';

describe(
	'StringInput',
	() => {
		afterEach(cleanup);

		it(
			'should render with type string',
			() => {
				const {asFragment} = render(
					<StringInput
						onChange={jest.fn()}
					/>
				);

				expect(asFragment()).toMatchSnapshot();
			}
		);

		it(
			'should render type string with value',
			() => {
				const mockOnChange = jest.fn();

				const {asFragment, getByTestId} = render(
					<StringInput
						onChange={mockOnChange}
						value={defaultValue}
					/>
				);

				expect(asFragment()).toMatchSnapshot();

				const element = getByTestId(SIMPLE_STRING_INPUT_TESTID);

				testControlledInput(
					{
						element,
						mockFunc: mockOnChange,
						value: defaultValue
					}
				);
			}
		);

		it(
			'should render type string with options',
			() => {
				const mockOnChange = jest.fn();

				const options = [
					{
						label: 'Default Value',
						value: 'defaultValue'
					}, {
						label: 'LIFERAY',
						value: 'Liferay'
					}
				];

				const {asFragment, getByTestId} = render(
					<StringInput
						onChange={mockOnChange}
						options={options}
						value={defaultValue}
					/>
				);

				expect(asFragment()).toMatchSnapshot();

				const element = getByTestId(OPTIONS_STRING_INPUT_TESTID);

				testControlledInput(
					{
						element,
						mockFunc: mockOnChange,
						value: defaultValue
					}
				);
			}
		);
	}
);