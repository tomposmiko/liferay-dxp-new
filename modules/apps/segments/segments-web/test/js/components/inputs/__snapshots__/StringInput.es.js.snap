// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StringInput should render type string with options 1`] = `
<DocumentFragment>
  <select
    class="form-control criterion-input form-control"
    data-testid="options-string"
  >
    <option
      value="defaultValue"
    >
      Default Value
    </option>
    <option
      value="Liferay"
    >
      LIFERAY
    </option>
  </select>
</DocumentFragment>
`;

exports[`StringInput should render type string with value 1`] = `
<DocumentFragment>
  <input
    class="criterion-input form-control"
    data-testid="simple-string"
    type="text"
    value="defaultValue"
  />
</DocumentFragment>
`;

exports[`StringInput should render with type string 1`] = `
<DocumentFragment>
  <input
    class="criterion-input form-control"
    data-testid="simple-string"
    type="text"
    value=""
  />
</DocumentFragment>
`;
